File: sub_pub/adapters/pulsar.py
"""Pulsar adapter"""
import logging
from typing import Iterator, List, Optional

from sub_pub.core.interfaces import MessageSource, MessagePublisher
from sub_pub.core.message import Message

logger = logging.getLogger(__name__)


class PulsarSource(MessageSource):
    """Pulsar message source"""
    
    def __init__(self, config: dict):
        self.config = config
        self.client = None
        self.consumer = None
        
    def connect(self) -> None:
        """Establish connection to Pulsar"""
        try:
            import pulsar
            
            service_url = self.config.pop('service_url')
            self.client = pulsar.Client(service_url)
            logger.info("Connected to Pulsar source")
        except ImportError:
            logger.error("pulsar-client package not installed. Install with: pip install pulsar-client")
            raise
        
    def subscribe(self, topics: List[str]) -> None:
        """Subscribe to topics"""
        if self.client:
            self.consumer = self.client.subscribe(
                topics,
                subscription_name=self.config.get('subscription_name', 'sub-pub'),
                **{k: v for k, v in self.config.items() if k != 'subscription_name'}
            )
            logger.info(f"Subscribed to Pulsar topics: {topics}")
        
    def consume(self) -> Iterator[Message]:
        """Consume messages from Pulsar"""
        if not self.consumer:
            raise RuntimeError("Consumer not connected")
        
        while True:
            msg = self.consumer.receive()
            
            headers = dict(msg.properties()) if msg.properties() else {}
            
            yield Message(
                payload=msg.data(),
                headers=headers,
                topic=msg.topic_name(),
                key=msg.partition_key() or None,
                timestamp=msg.publish_timestamp()
            )
        
    def close(self) -> None:
        """Close Pulsar connection"""
        if self.consumer:
            self.consumer.close()
        if self.client:
            self.client.close()
            logger.info("Pulsar source closed")
        
    def commit(self, message: Optional[Message] = None) -> None:
        """Acknowledge message"""
        # In Pulsar, we'd need to store the actual pulsar message to ack it
        # For now, this is a simplified version
        pass


class PulsarPublisher(MessagePublisher):
    """Pulsar message publisher"""
    
    def __init__(self, config: dict):
        self.config = config
        self.client = None
        self.producers = {}  # topic -> producer
        
    def connect(self) -> None:
        """Establish connection to Pulsar"""
        try:
            import pulsar
            
            service_url = self.config.pop('service_url')
            self.client = pulsar.Client(service_url)
            logger.info("Connected to Pulsar publisher")
        except ImportError:
            logger.error("pulsar-client package not installed. Install with: pip install pulsar-client")
            raise
        
    def publish(self, message: Message, topic: str) -> None:
        """Publish a message to Pulsar"""
        if not self.client:
            raise RuntimeError("Client not connected")
        
        # Create producer for topic if not exists
        if topic not in self.producers:
            self.producers[topic] = self.client.create_producer(topic)
        
        producer = self.producers[topic]
        
        producer.send(
            message.payload,
            properties=message.headers,
            partition_key=message.key
        )
        
    def flush(self) -> None:
        """Flush buffered messages"""
        for producer in self.producers.values():
            producer.flush()
        
    def close(self) -> None:
        """Close Pulsar connection"""
        for producer in self.producers.values():
            producer.close()
        if self.client:
            self.client.close()
            logger.info("Pulsar publisher closed")

File: rust/src/adapters/pulsar.rs
use async_trait::async_trait;

use crate::core::{Message, MessagePublisher, MessageSource};
use crate::error::SubPubError;

/// Stub Apache Pulsar source.
pub struct PulsarSource {
    pub service_url: String,
    pub subscription: String,
    pub topics: Vec<String>,
}

impl PulsarSource {
    pub fn new(service_url: impl Into<String>, subscription: impl Into<String>) -> Self {
        Self {
            service_url: service_url.into(),
            subscription: subscription.into(),
            topics: Vec::new(),
        }
    }
}

#[async_trait]
impl MessageSource for PulsarSource {
    async fn connect(&mut self) -> Result<(), SubPubError> {
        tracing::warn!("PulsarSource: full implementation pending");
        Ok(())
    }

    async fn subscribe(&mut self, topics: &[&str]) -> Result<(), SubPubError> {
        self.topics = topics.iter().map(|t| t.to_string()).collect();
        Ok(())
    }

    async fn consume(&mut self) -> Result<Option<Message>, SubPubError> {
        Err(SubPubError::Adapter(
            "PulsarSource: full implementation pending".to_string(),
        ))
    }

    async fn commit(&mut self, _message: &Message) -> Result<(), SubPubError> {
        Ok(())
    }

    async fn close(&mut self) -> Result<(), SubPubError> {
        Ok(())
    }
}

/// Stub Apache Pulsar publisher.
pub struct PulsarPublisher {
    pub service_url: String,
}

impl PulsarPublisher {
    pub fn new(service_url: impl Into<String>) -> Self {
        Self {
            service_url: service_url.into(),
        }
    }
}

#[async_trait]
impl MessagePublisher for PulsarPublisher {
    async fn connect(&mut self) -> Result<(), SubPubError> {
        tracing::warn!("PulsarPublisher: full implementation pending");
        Ok(())
    }

    async fn publish(&mut self, _message: &Message, _topic: &str) -> Result<(), SubPubError> {
        Err(SubPubError::Adapter(
            "PulsarPublisher: full implementation pending".to_string(),
        ))
    }

    async fn flush(&mut self) -> Result<(), SubPubError> {
        Ok(())
    }

    async fn close(&mut self) -> Result<(), SubPubError> {
        Ok(())
    }
}
